### ISSUE TITLE
Problem about accuracy

### ISSUE THREAD

**Issue by bwmaples** [Thijsvanede/FlowPrint#4]
```python
# trainSet: the path of trainSet (dir of train pcap files)
# trainLabel: the package name of traffic
# testSet the path of testSet (dir of test pcap files)
def flowprintTest(trainSet, trainLabel, testSet):
    preprocessor = Preprocessor(verbose=True)
    if(os.path.exists("./flows.p")):
        X_train, y_train = preprocessor.load("./flows.p")
    else:
        X_train, y_train = preprocessor.process(files=trainSet,
                                labels=trainLabel)
        preprocessor.save('flows.p', X_train, y_train)

    flowprint = FlowPrint(
        batch       = 300,
        window      = 30,
        correlation = 0.1,
        similarity  = 0.9
    )
    flowprint.fit(X_train, y_train)

    fingerprints = flowprint.fingerprints
    with open("./fingerprints","w") as fp:
        for fingerprint in fingerprints:
            fp.write(fingerprints[fingerprint] + " " + json.dumps(fingerprint.to_dict())+"\n")

    totalNum = 0.0
    trueNum = 0.0
    for testData in testSet:
        fileName = testData.split("\\")[-1]
        fileName = fileName.split(".pcap")[0]
        # Get X_test fingerprints
        X_test = np.array(list(preprocessor.extract(testData).values()))
        # Is this function?
        testPrints = flowprint.fingerprinter.fit_predict(X_test)
        y_recognize = flowprint.recognize(testPrints)
        if(fileName in y_recognize.tolist()):
            trueNum += 1
        else:
            with open("./result.txt","a") as fp:
                fp.write(fileName+"\n")
                if(np.size(y_recognize)==0):
                    fp.write("[]"+"\n")
                else:
                    fp.write(y_recognize[0]+"\n")
                for testPrint in testPrints.tolist():
                    fp.write(json.dumps(testPrint.to_dict())+"\n")
                fp.write("############\n")
        totalNum += 1

    if(totalNum!=0):
        Trate = trueNum/totalNum
        print("Success rate is " + str(Trate))
    return

if __name__ == "__main__":
    fileNames = os.listdir(trainPath)
    fileNames.sort()
    labels = []
    inputFlows = []
    testFlows = []
    for fileName in fileNames:
        labels.append(fileName.split(".pcap")[0])
        inputFlows.append(trainPath+fileName)

    fileNames = os.listdir(testPath)
    for fileName in fileNames:
        testFlows.append(testPath+fileName)

    flowprintTest(inputFlows,labels,testFlows)
```
I'm using this function to reproduce the experiment.
But I can only get 50% accuracy.
The dataSet I am using is that https://recon.meddle.mobi/cross-market.html -China-android.
I split the data set in half, half as the training set and half the test set.
Is the way I use it wrong?

**Comment by Thijsvanede** [comment]
Hi bwmaples,

I have taken a quick look at the code (also the parts that you sent to me via e-mail) and the main difference I noticed was in the way we split the data:

In our FlowPrint work we split the data into training and testing sets based on the flows, instead of packets. This was done to make sure both training and testing data have enough flows to train and test with instead of e.g., one single big flow with a lot of packets. Please see the `fingerprint()` function in `__main__.py`.

I have added a script under `/examples/recognition.py` that performs the experiment, hopefully this helps with the reproduction. Note that this script is not yet available in the pip release. Using the input
```
python3 recognition.py --dir </path/to/cross_market>/china/android/ --ratio 0.5 --random
```

I get the following results:
```
              precision    recall  f1-score   support
[...]
    accuracy                         0.8890     26127
   macro avg     0.9218    0.9133    0.9105     26127
weighted avg     0.9187    0.8890    0.8929     26127
```

In addition we used a random split to get a full coverage of the use of an app for both training and testing while making sure the flows do not overlap. We can also split the flows 50:50 as in your code, the problem then is that different functionality or modules of the app are present in training and testing. This should make both AppScanner and FlowPrint produce worse results. If you'd like to test with that remove the `--random` parameter in the previously mentioned script

Please let me know if this helps

**Comment by bwmaples** [comment]
Thank you very much for your reply! It really worked out!
And random is necessary.
I got a similar weighted avg to yours.
Really a meaningful research job!
